### ISSUE TITLE
NPE while using by inject, due to lazy initialization in JobIntentService(koin v2.1.5)

### ISSUE THREAD

**Issue by adiamartya** [InsertKoinIO/koin#797]
I am using joinAll inside a jobintentservice. Also, i m initializing global objects via inject. This was working fine on v2.0.1. When i upgraded it to v2.1.5, two crashes have been logged in crashlytics

Fatal Exception: kotlin.KotlinNullPointerException
       at kotlin.UnsafeLazyImpl.getValue(UnsafeLazyImpl.java:81)
       at com.android.david.onboarding.services.SuggestionsService.getSplashRepo(SuggestionsService.java:1)
       at com.android.david.onboarding.services.SuggestionsService.access$getSplashRepo$p(SuggestionsService.java:16)
       at com.android.david.onboarding.services.SuggestionsService$onHandleWork$1$2.invokeSuspend(SuggestionsService.java:41)
       at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(BaseContinuationImpl.java:33)
       at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.java:56)
       at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.java:571)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.java:738)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.java:678)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.java:665)

**Comment by lateplate** [comment]
Seeing something similar with 2.1.5, though I'm not entirely sure if it's Koin or something else. We are doing a fairly straightforward `inject` in a class that extends `KoinComponent`, and for some reason we are hitting an NPE when trying to inject that object. 

Crashes seem to be random, especially for a class that is _constantly_ in use (used in every Retrofit/API call we make). If it were something more deep rooted, I would think we would see thousands of crashes, but only a couple dozen here.

Stack:
```kotlin
kotlin.KotlinNullPointerException: null
    at kotlin.UnsafeLazyImpl.getValue(Lazy.kt:81)
    at com.basecamp.hey.api.AuthInterceptor.getSession
    at com.basecamp.hey.api.AuthInterceptor.setRefreshInProgress
    at com.basecamp.hey.api.AuthInterceptor.getSession(AuthInterceptor.kt:0)
    at com.basecamp.hey.api.AuthInterceptor.getResponse(AuthInterceptor.kt:60)
    at com.basecamp.hey.api.AuthInterceptor.intercept(AuthInterceptor.kt:32)
    at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:100)
    at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:197)
    at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:502)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
    at java.lang.Thread.run(Thread.java:919)
```

Line 60 of `AuthInterceptor` is a reference to a `Session` object that is created like so:

```kotlin
    private val session by inject<Session>()
```

Any help is appreciated!

**Comment by arnaudgiuliani** [comment]
do you have more stack?

**Comment by RamiTrabelsiAndroid** [comment]
We're encountering a very similar problem, just after upgrading `Koin` from `v1.0.2` to `v2.1.6`, new crashes has popped up on our `Crashlytics`. Crashes seems to be at the lines using `inject() ` like this : 

`private val database: AppDatabase? by inject()`

The 2 stack traces that we have, are happening when injecting databases or repositories with Dao classes (constructor injected) -don't know if it's coincidence for this-. 

We're using Room, I paste 3 crashs stack traces below : 

`
Fatal Exception: kotlin.KotlinNullPointerException

kotlin.UnsafeLazyImpl.getValue (UnsafeLazyImpl.java:81)

a.b.c.d$XXX$.getDatabase (XXX.java:3)

a.b.c.d$XXX$p (XXX.java:23)

a.b.c.d$XXX$method$1.invokeSuspend (XXX.java:122)

a.b.c.d$XXX$$loadFromDb$1.invoke (XXX.java:10)
`

`
Fatal Exception: kotlin.KotlinNullPointerException

kotlin.UnsafeLazyImpl.getValue (UnsafeLazyImpl.java:81)

a.b.c.d$XXX$.getUserSessionRepository (SessionManager.java:2)

a.b.c.d$XXX$.restoreSessionSync (SessionManager.java:158)

a.b.c.d$XXX$restoreSession$2.invokeSuspend (SessionManager.java:137)

a.b.c.d$XXX$restoreSession$2.invoke (SessionManager.java:10)
`

Thank you in advance for taking a look at this serious problem affecting big number of our users.

**Comment by lateplate** [comment]
> do you have more stack?

@arnaudgiuliani That's about all we have. It's coming from a crash log and I can't find a way to locally reproduce it where I might be able to grab more logs.

**Comment by shaileshmota** [comment]
@arnaudgiuliani for the last trace :  https://github.com/InsertKoinIO/koin/issues/797#issuecomment-665149641

We are creating a module for Room DB like : 
```
val databaseModule = module {

    single {
        Room.databaseBuilder(get(), MyDb::class.java, "my_db")
            .fallbackToDestructiveMigration()
            .build()
    }
}
```

Further in a ViewModel we inject it like : 
```
class MyViewModel : ViewModel(), KoinComponent {

    private val viewModelJob = Job()

    private val viewModelScope = CoroutineScope(Dispatchers.Main + viewModelJob)

    private val database by inject<MyDb>()
    .
    .
    .
```
It seems like further accessing the instance `database` produces NPE

**Comment by shaileshmota** [comment]
This crash is quite considerable on our crash stats. Due to lack of activity on this thread we will simply downgrade the version to see where it works.  @lateplate can you tell us from what version did you upgrade to 2.1.5 ? 
It seems like it is working fine for @adiamartya for version 2.0.1

**Comment by lateplate** [comment]
@shaileshmota We kept pretty up to date, so I believe we were on 2.1.4 previously. But do note that our crashes are pretty sporadic and not super frequent.

**Comment by RamiTrabelsiAndroid** [comment]
Hello @lateplate, you already had crashes with version 2.1.4 or everything was fine with that version ?

**Comment by lateplate** [comment]
Couldn’t say for sure, unfortunately. Our crash data is pretty limited so I can’t say when it may have started happening. 

notifications@github.com wrote:
“Hello @lateplate, you already had crashes with version 2.1.4 or everything was fine with that version ?”

**Comment by ilya-shknaj** [comment]
After update from **2.0.0** to **2.1.5** we see a lot of crashes in production.
During last 30 days ~ 5000 crashes. Unfortunately, I don't have steps for reproducing.
<img width="1286" alt="Screen Shot 2020-09-25 at 10 11 06" src="https://user-images.githubusercontent.com/4130507/94237392-7629c100-ff17-11ea-8ade-b9d5aec3a2f6.png">

But, I've reviewed code and found a critical change that possibly leads to this issue
https://github.com/InsertKoinIO/koin/commit/45ce7d240981080ce2ba88b5d3e88946b7fbead7
It appeared in 2.1.0-aplha9
@lateplate , @arnaudgiuliani Could you check that?

**Comment by egorikftp** [comment]
@arnaudgiuliani Any updates?

**Comment by arnaudgiuliani** [comment]
I need your stacktrace and code usage. Hard to say like that :/

**Comment by egorikftp** [comment]
@arnaudgiuliani The stacktrace the same as mentions above in the issue. We have made the research and found that one of the reason of this bug is making  **lazy(LazyThreadSafetyMode.NONE)**. 

Could you please explain the reason of this change?

**Comment by arnaudgiuliani** [comment]
`lazy(LazyThreadSafetyMode.NONE)` optimisation proposal I guess. 🤔 

It's something I can look at.

**Comment by Qoxis** [comment]
We have the same errors since we use Koin to init Room and our DAOs, It's now the first error in crashlytics.
Here is the init code:

```
            modules(
                module {
                    single {
                        Room.databaseBuilder(
                            androidApplication(),
                            AppDatabase::class.java,
                            "my-db"
                        ).fallbackToDestructiveMigration().build()
                    }
                    single { get<AppDatabase>().offerDao() }
                    single { get<AppDatabase>().userDao() }
                    single { get<AppDatabase>().websitesDao() }
                }
            ) 
```

And we get;

Fatal Exception: java.lang.NullPointerException

> kotlin.UnsafeLazyImpl.getValue (UnsafeLazyImpl.java:81)
> getOfferDao (OffersRepository.java:2)
> OffersRepository.access$getOfferDao$p (OffersRepository.java:14)
> OffersRepository$metaMosaic$1.invoke (OffersRepository.java:45)
> OffersRepository$metaMosaic$1.invoke (OfferscRepository.java:14)
> ResultKt$resultLiveData$1.invokeSuspend (ResultKt.java:26)
> kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith (BaseContinuationImpl.java:33)
> kotlinx.coroutines.DispatchedTask.run (DispatchedTask.java:56)
> kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely (CoroutineScheduler.java:571)
> kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask (CoroutineScheduler.java:738)
> kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker (CoroutineScheduler.java:678)
> kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run (CoroutineScheduler.java:665)

Looking forward an update on this error, thanks :)

**Comment by arnaudgiuliani** [comment]
I will push a rc-4 with my last fix, can anyone test it?

**Comment by Qoxis** [comment]
Yes we can make an internal build and then deploy it to 5% of our base and check if it still happen.

**Comment by arnaudgiuliani** [comment]
rc-4 is online!

**Comment by trashkalmar** [comment]
Please, consider adding extensions say `threadUnsafeInject()` for using inside UI, when it's known that there will never be threading problems, so removing synchronization overhead.

**Comment by arnaudgiuliani** [comment]
to people that have lazy evaluation problem, is it mostly located in Android `Service` components?

Not sure, it's super impecting for now to have standard `lazy` resolution. But finally, it seems to be there only on Service. Then I would propose to come back to Lazy(NONE) on UI extensions. Service would keep standard Lazy evaluation.

**Comment by arnaudgiuliani** [comment]
The trade-off I can propose is :
- have Lazy(None) on Android UI extensions (inject, viewModel ...)
- keep default Lazy strategy for everything else as default (Service, Worker, KoinComponent ...)

What do you think?

**Comment by trashkalmar** [comment]
It may be confusing to have the tool that works different in different places. Actually, it'd be nice to have a manual control on the threading while injecting anyway. For my projects I've made two extension methods for this purpose, so I'm sure about the threading:

```
// App's Koin
@OptIn(KoinApiExtension::class)
interface AppKoin : KoinComponent {
  override fun getKoin() = ...
}

// Manual threading control with NONE as default
inline fun <reified T : Any> AppKoin.inject(qualifier: Qualifier? = null, mode: LazyThreadSafetyMode = LazyThreadSafetyMode.NONE) =
    lazy(mode) {
      getKoin().get<T>(qualifier)
    }

// Thread-safe injection
inline fun <reified T : Any> AppKoin.safeInject(qualifier: Qualifier? = null) =
    inject<T>(qualifier, LazyThreadSafetyMode.SYNCHRONIZED)

```

**Comment by arnaudgiuliani** [comment]
If we check Android sources, they don't use Lazy<T> - https://cs.android.com/androidx/platform/frameworks/support/+/androidx-master-dev:lifecycle/lifecycle-viewmodel-ktx/src/main/java/androidx/lifecycle/ViewModelProvider.kt;l=41?q=%20ViewModelLazy

**Comment by arnaudgiuliani** [comment]
closing this issue for now. Let's keep in touch for better Lazy implementation for UI.

**Comment by Qoxis** [comment]
On our side, no more errors since the rc-4. Thanks again !

**Comment by vegeta2102** [comment]
@arnaudgiuliani 
This problem solved on`2.2.0-rc-4` . But It seems occur again in `2.2.0`
